ranger/core/actions.py

```diff
@@ -49,6 +49,8 @@ class _MacroTemplate(string.Template):
 class Actions(  # pylint: disable=too-many-instance-attributes,too-many-public-methods
         FileManagerAware, SettingsAware):
 
+    _macros = None
+    _macro_reg = re.compile(ranger.MACRO_DELIMITER + r"\w*", re.IGNORECASE)
```

Comment:
Why do we need this regex now?

Reply:
Because we want to extract all macros from the given string. For example, if the string to substitute macros in is "shell echo %confdir > %f", we would like to extract "%confdir" and "%f" and generate the macro expansions only for those.

Reply:
Oh, because previously whatever the substitute needed would be present anyway? I think this'll lead to `KeyError`s. Maybe we should check whether each key is in the string? Though I'm not sure that's possible for the variable keys like `%5d`.

Reply:
Yes, the previous macro system was generating the whole dict each time the macros of a string were being expanded. If there is a macro in the string that ranger does not define, it just won't get expanded (it was so and now it is also so). No `KeyError`s will be generated so I don't quite understand what you mean.

Reply:
Oh, I'd missed the if test. I know the regex is more efficient but are we sure we don't want other characters in future macros, and does it catch *all* the edge cases? Going through all possible macros every time, checking if they're a substring is terrible too tbh especially because macros could be substrings of other macros : /

I think the most pythonic way might be to override the get methods of the dict, through subclassing for example. [As noted here](https://stackoverflow.com/q/3387691) we'd have several methods to override — off the top of my head there's `__getitem__, get, pop` and maybe `_process_args` — but it'd be the most transparent solution I can think off. No values are unnecessarily evaluated and no fragile regex to determine the macros we really need.

@Vifon, since this is quite some work, what's your input on this?

Reply:
If you would like to, I can make the regex a little bit more complicated and it will catch everything. For example
```
%[^(?!%\s)]*
```
matches "%" then everything that is not "%s" and whitespace. It essentially matches up to another "%" or space.

Reply:
The problem is I don't trust anyone, not even myself to come up with a regex that'll catch everything.